Automattic/wp-calypso
# Editor: Add PostListStoreFactory

Part of #303 - In order to have multiple components in a given view that need to interact with distinct query options of a `post-list-store` - the current `post-list-store` needs to be abstracted into a factory akin to the `category-store`.

__Background__
While building the "Parent Page Selector" in the editor - both the parent page selector and the post scheduler were modifying the `post-list-store` query options, and as such the query results.  This lead to some odd behavior between the two, and ultimately was fixed by disabling the post scheduler when editing pages.

By moving to a factory setup, components like the `post-list-fetcher` can create their own instance of a `post-list-store` to ensure query options and results are not polluted by other components on the page.

This branch indeed got a bit larger then I typically like, but a large amount of the new lines are in the form of tests.  Additionally, reading the diff of `post-list-store` might be a bit difficult and is probably best read outside of the diff viewer.  I did not change any logic within the `post-list-store`, just restructured it to flow with the factory setup, and ES6-ified it.

__To Test__
Test coverage for the `post-list-store` has been added in this branch.  The tests can be executed by performing `cd client/lib/posts && make`.

Additionally, the `post-list-store` acts as a data provider for the following components.  All should be interacted with during testing to ensure no odd behavior is encountered

- `my-sites/drafts/draft-list.jsx` Can be tested by using the "Drafts Drawer" in the editor
- `my-sites/menus/item-options/post-list.jsx` Test via the menu builder by adding menu options for both a page and a post.  Interact with the search box here as well and validate the search filter still operates as expected
- `my-sites/pages/page-list.jsx` View a page list for a site.  Interact with the filter bar by toggling between Drafts and Trash.  Filter pages using search
- `my-sites/post-selector/index.jsx` Open a page in the editor, interact with the parent page selector in the sidebar.
- `my-sites/posts/post-list.jsx` Open a sites posts list.  Interact with the filterbar and search.
- `post-editor/editor-ground-control/index.jsx` Inside the post editor, expand the post scheduler.  Ensure calendar items are circled as expected for dates where posts have been published.
- Stats Insights Page - The "Latest Post Summary" on the insights tab of the stats page also uses the `post-list-store`.  Validate that data shows up as expected inside this module
